core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java

```diff
@@ -574,14 +575,27 @@ private void writeIntrospectionClass(ClassWriterOutputVisitor classWriterOutputV
         dispatchWriter.buildGetTargetMethodByIndex(classWriter);
         buildFindIndexedProperty(classWriter);
         buildGetIndexedProperties(classWriter);
-
+        boolean hasBuilder = annotationMetadata != null && annotationMetadata.isPresent(Introspected.class, "builder");
         if (defaultConstructor != null) {
             writeInstantiateMethod(classWriter, defaultConstructor, "instantiate");
+            // in case invoked directly or via instantiateUnsafe
+            if (constructor == null) {
+                writeInstantiateMethod(classWriter, defaultConstructor, "instantiateInternal", Object[].class);
+                writeBooleanMethod(classWriter, METHOD_IS_BUILDABLE, true);
+            }
         }
-        if (constructor != null && ArrayUtils.isNotEmpty(constructor.getParameters())) {
+
+        if (constructor != null) {
+            if (ArrayUtils.isEmpty(constructor.getParameters())) {
+                writeInstantiateMethod(classWriter, constructor, "instantiate");
+            }
             writeInstantiateMethod(classWriter, constructor, "instantiateInternal", Object[].class);
+            writeBooleanMethod(classWriter, METHOD_IS_BUILDABLE, true);
+        } else if (defaultConstructor == null) {
+            writeBooleanMethod(classWriter, METHOD_IS_BUILDABLE, hasBuilder);
```

Comment:
Would be cool if we could also write the builder as a constructor (write the `instantiate` method or another method). This way serde and other modules that use the API would need less changes. It could also eliminate some overhead that is in the `BeanIntrospection.builder`. But I guess it would be hard and we would need to decide which one to write if there is both builder and constructor.

Reply:
so I implemented this because it is probably useful for compatibility however, I believe that we will want to optimise this to use the `builder()` API in serde because otherwise the JSON will need to be buffered // cc @yawkat